lib/cylc/remote.py

```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+
+    def process_get_host_metric_cmd(self):
+        """Return the command string to run 'cylc get host metric' with (only)
+           the required options given rank method and thresholds specified."""
+        cmd = self.CMD_BASE
+        if self.RANK_METHOD != 'random':
+            cmd = " ".join((cmd, self.opt_mapping[self.RANK_METHOD][0]))
+        if self.PARSED_THRESHOLDS:
+            for measure in self.PARSED_THRESHOLDS.keys():
+                if measure != self.RANK_METHOD:
+                    cmd = " ".join((cmd, self.opt_mapping[measure][0]))
+        return cmd
+
+    def remove_bad_hosts(self, cmd_with_opts, mock_stats=False):
+        """Run 'get-host-metric' on each host & store extracted stats for
+           'good' hosts only. Ignore 'bad' hosts whereby either metric data
+           cannot be accessed from the command or at least one metric value
+           does not pass a specified threshold."""
+        host_stats = {}
+        if mock_stats:  # Create fake data for unittest purposes (only).
+            host_stats = mock_stats
+        else:
+            for host in self.HOSTS:
+                if host == 'localhost':
+                    process = remote_cylc_cmd(cmd_with_opts.split(),
+                                              capture=True)
+                else:
+                    process = remote_cylc_cmd(cmd_with_opts.split(),
+                                              capture=True, host=host)
+                metric = process.communicate()[0]
+                process.wait()
+                ret_code = process.returncode
+                if ret_code:
+                    # Can't access data => designate as 'bad' host & ignore.
+                    print "Can't obtain metric data from host '%s'." % host
+                    raise CalledProcessError(ret_code, cmd_with_opts.split())
+                else:
+                    host_stats[host] = json.loads(metric)
+
+        bad_hosts = []  # Get errors if alter dict during iteration. Use list.
+        for host in host_stats:
+            for measure, cutoff in self.PARSED_THRESHOLDS.items():
+                datum = self.access_datum(host_stats[host],
+                                          self.opt_mapping[measure][1])
+                # Cutoff is a minimum or maximum depending on measure context.
+                if ((measure in ("memory", "disk-space") and
+                    datum < cutoff) or
+                    (measure in ("load:1", "load:5", "load:15")
+                     and datum > cutoff)):
+                    bad_hosts.append(host)
+                    continue
+        return dict((host, metr) for host, metr in
+                    host_stats.items() if host not in bad_hosts)
+
+    def rank_good_hosts(self, all_host_stats):
+        """Take a dictionary of hosts and corresponding metric data and rank
+           hosts via the specified rank method, returning the most suitable."""
+        hosts_with_values_to_rank = dict((
```

Comment:
I'll check - it certainly seems plausible & would be clearer.

Reply:
On closer inspection, the ``rank_good_hosts`` code (at this review stage) was in an utterly incomprehensible format & generally hideous, so apologies for that! I have played around with the expressions there & can't seem to get something similar to your snippet to work. Notably, I need to retain the ranked hosts, not the ranked metrics, so it can't be quite as simple as this.

I have, however managed to completely untangle ``sort_asc_hosts`` which was the main offender in terms of comprehensibility & in combination with some piecewise comments I have added, I think this is now sufficiently readable (see latest commits). But feel free to disagree!

NB: I did find a few fancy solutions using ``operator`` module functions as ``keys`` for ``sorted()`` but they seemed a bit far-fetched.